ext/random/random.c

```diff
@@ -85,14 +85,18 @@ static zend_object_handlers random_randomizer_object_handlers;
 static inline uint32_t rand_range32(const php_random_algo *algo, php_random_status *status, uint32_t umax)
 {
 	uint32_t result, limit, r;
-	size_t total_size = 0;
+	size_t total_size = 0, shift_size = 0;
 	uint32_t count = 0;
 
 	result = 0;
 	total_size = 0;
 	do {
 		r = algo->generate(status);
-		result = (result << (8 * status->last_generated_size)) | r;
+		shift_size = (8 * status->last_generated_size);
```

Comment:
Note that my previous comment was wrong, it looks like gcc and clang both treat `x << y` as `x << (y % 32)`, so this is a behavior change. See:

https://godbolt.org/z/W1xn847Eh

This would be the equivalent here:

```c
shift_size = (8 * status->last_generated_size) % (8 * sizeof(result));
```

Reply:
I was about to say the same 😄 https://www.ideone.com/ijVb65
So I think you should revert the changes, then just change <code>status->last_generated_size</code> to <code><b>(</b>status->last_generated_size<b> % sizeof(result))</b></code> in the shift?

Reply:
Indeed! It's so complicated 😅 
fix again: 2ebbefc94b62f9e608a41331c58adfc51cbb18f0

Reply:
Ah I hadn't seen Ilija's main comment:

> The change in behavior might or might not matter, I don't know much about the logic of this function.

Indeed, maybe here you would prefer to handle
`result << (8 * status->last_generated_size)` not as
`result << (8 * (status->last_generated_size % sizeof(result)))` but rather as
`status->last_generated_size >= sizeof(result) ? 0 : result << (8 * status->last_generated_size)`? https://www.ideone.com/Oy1PqS

Reply:
Additional thought: whatever the solution, it should be applied to both `rand_range32` and `rand_range64`.
I see that #9088 does, and also tackles an endianness issue 🙂